src/preprocessing/particle_packing/nhs_faces.jl

```diff
@@ -0,0 +1,263 @@
+    return triangle_plane_intersection(v1, normal, min_corner, max_corner, cell_size)
+end
+
+# See https://tavianator.com/2022/ray_box_boundary.html
+function ray_intersection(min_corner, max_corner, ray_origin, ray_direction;
+                          pad=sqrt(eps()))
+    NDIMS = length(ray_origin)
+
+    inv_dir = SVector(ntuple(@inline(dim->1 / ray_direction[dim]), NDIMS))
+
+    tmin = zero(eltype(ray_direction))
+    tmax = Inf
+    @inbounds for dim in 1:NDIMS
+        # `pad` is to handle rays on the boundary
+        t1 = (min_corner[dim] - pad - ray_origin[dim]) * inv_dir[dim]
+        t2 = (max_corner[dim] + pad - ray_origin[dim]) * inv_dir[dim]
+
+        tmin = min(max(t1, tmin), max(t2, tmin))
+        tmax = max(min(t1, tmax), min(t2, tmax))
+    end
+
+    return tmin <= tmax
+end
+
+# Check if each cell vertex is located on the same side of the plane.
+# Otherwise the plane intersects the cell.
+function triangle_plane_intersection(point, plane_normal, min_corner, max_corner, cell_size)
+    dirx = SVector(cell_size[1], zero(eltype(point)), zero(eltype(point)))
+    diry = SVector(zero(eltype(point)), cell_size[2], zero(eltype(point)))
+    dirz = SVector(zero(eltype(point)), zero(eltype(point)), cell_size[3])
+
+    cell_vertex_1 = min_corner
+    cell_vertex_2 = min_corner + dirx
+
+    pos_diff_1 = cell_vertex_1 - point
+    pos_diff_2 = cell_vertex_2 - point
+
+    # Corners: bottom north-west and bottom north-east
+    dot1 = sign(dot(pos_diff_1, plane_normal))
+    dot2 = sign(dot(pos_diff_2, plane_normal))
+    !(dot1 == dot2) && return true
+
+    cell_vertex_3 = min_corner + diry
+    pos_diff_3 = cell_vertex_3 - point
+
+    # Corners: bottom north-east and top north-west
+    dot3 = sign(dot(pos_diff_3, plane_normal))
+    !(dot2 == dot3) && return true
+
+    cell_vertex_4 = min_corner + dirz
+    pos_diff_4 = cell_vertex_4 - point
+
+    # Corners: top norht-west and bottom south-west
+    dot4 = sign(dot(pos_diff_4, plane_normal))
+    !(dot3 == dot4) && return true
+
+    cell_vertex_5 = max_corner
+    pos_diff_5 = cell_vertex_5 - point
+
+    # Corners: bottom south-west and top south-east
+    dot5 = sign(dot(pos_diff_5, plane_normal))
+    !(dot4 == dot5) && return true
+
+    cell_vertex_6 = max_corner - dirx
+    pos_diff_6 = cell_vertex_6 - point
+
+    # Corners: top south-east and top south-west
+    dot6 = sign(dot(pos_diff_6, plane_normal))
+    !(dot5 == dot6) && return true
+
+    cell_vertex_7 = max_corner - diry
+    pos_diff_7 = cell_vertex_7 - point
+
+    # Corners: top south-west and bottom south-east
+    dot7 = sign(dot(pos_diff_7, plane_normal))
+    !(dot6 == dot7) && return true
+
+    cell_vertex_8 = max_corner - dirz
+    pos_diff_8 = cell_vertex_8 - point
+
+    # Corners: bottom south-east and top north-east
+    dot8 = sign(dot(pos_diff_8, plane_normal))
+    !(dot7 == dot8) && return true
+
+    # All edge vertices are on one side of the plane
+    return false
+end
```

Comment:
```suggestion
function triangle_plane_intersection(point, plane_normal, min_corner, max_corner, cell_size)
    # Generate all eight corners of the cell (axis-aligned bounding box).
    # Each corner is calculated by adding a combination of cell_size components
    # to the min_corner, based on the indices i, j, k, which take values 0 or 1.
    # This creates all possible combinations, resulting in the eight corners of the cell.
    corners = [
        min_corner + SVector(i * cell_size[1], j * cell_size[2], k * cell_size[3])
        for i in (0, 1), j in (0, 1), k in (0, 1)
    ]

    # For each corner, compute the vector from a known point on the plane to the corner.
    # Then, calculate the dot product of this vector with the plane's normal vector.
    # The dot product indicates the relative position of the corner to the plane:
    # - Positive value: corner is on the same side as the normal vector points.
    # - Zero: corner lies exactly on the plane.
    # - Negative value: corner is on the opposite side from where the normal vector points.
    # The sign() function converts the dot product into -1, 0, or 1 accordingly.
    signs = [sign(dot(corner - point, plane_normal)) for corner in corners]

    # Determine if the plane intersects the cell by checking the signs.
    # - If all signs are the same (all corners on one side of the plane), there's no intersection.
    # - If signs vary (some corners on different sides), the plane intersects the cell.
    # If maximum(signs) > minimum(signs), it means there are both positive and negative signs.
    return maximum(signs) > minimum(signs)
end

```

Reply:
Thanks for this. Your version is definitely more readable. I'm wondering if my version is slightly faster as we don't have to check each combination in best case.

However, I prefer your readable version. What's @efaulhaber opinion on this?

Reply:
Please just test what the performance is like compared to the current version.

Reply:
That's not that easy. I mean it's case dependent. In best case (first condition is true), it should be faster. But in worst case it should be same performance, no?

Reply:
Yes but I would just compare it for a large example and check if the impact is actually measurable.

Reply:
I think the performance difference is negligible, that's why I also prefer your version.

Reply:
```julia
function triangle_plane_intersection(point_on_plane, plane_normal, cell_min_corner, cell_size)
    cell_center = cell_min_corner .+ cell_size ./ 2

    # The corner that is furthest in the direction of `plane_normal`.
    # When the plane is not intersecting the cell, this corner is the one closest
    # to the plane when the normal is pointing away from the cell or furthest from the plane
    # when the normal is pointing towards the cell.
    # Note that this could also be a face or edge midpoint when the plane is axis-aligned.
    normal_unit = sign.(plane_normal)
    corner1 = cell_center + normal_unit .* cell_size / 2
    corner2 = cell_center - normal_unit .* cell_size / 2

    # These two vectors are on the same side of the plane
    # if and only if the plane intersects the cell.
    plane_to_corner1 = corner1 - point_on_plane
    plane_to_corner2 = corner2 - point_on_plane

    # Return true if the two vectors are on different sides of the plane
    return dot(plane_normal, plane_to_corner1) * dot(plane_normal, plane_to_corner2) < 0
end
```

Tested for random samples:
```
Default:
  7.340 ns (0 allocations: 0 bytes)
Sven:
  63.733 ns (4 allocations: 416 bytes)
Erik:
  4.791 ns (0 allocations: 0 bytes)

Default:
  7.333 ns (0 allocations: 0 bytes)
Sven:
  107.200 ns (4 allocations: 416 bytes)
Erik:
  4.833 ns (0 allocations: 0 bytes)

Default:
  7.333 ns (0 allocations: 0 bytes)
Sven:
  114.166 ns (4 allocations: 416 bytes)
Erik:
  4.791 ns (0 allocations: 0 bytes)

Default:
  7.340 ns (0 allocations: 0 bytes)
Sven:
  75.069 ns (4 allocations: 416 bytes)
Erik:
  4.833 ns (0 allocations: 0 bytes)

Default:
  7.333 ns (0 allocations: 0 bytes)
Sven:
  111.201 ns (4 allocations: 416 bytes)
Erik:
  4.833 ns (0 allocations: 0 bytes)

Default:
  7.333 ns (0 allocations: 0 bytes)
Sven:
  72.371 ns (4 allocations: 416 bytes)
Erik:
  4.791 ns (0 allocations: 0 bytes)

Default:
  3.666 ns (0 allocations: 0 bytes)
Sven:
  115.297 ns (4 allocations: 416 bytes)
Erik:
  4.833 ns (0 allocations: 0 bytes)

Default:
  7.333 ns (0 allocations: 0 bytes)
Sven:
  74.966 ns (4 allocations: 416 bytes)
Erik:
  4.833 ns (0 allocations: 0 bytes)

Default:
  7.375 ns (0 allocations: 0 bytes)
Sven:
  116.703 ns (4 allocations: 416 bytes)
Erik:
  4.833 ns (0 allocations: 0 bytes)

Default:
  7.333 ns (0 allocations: 0 bytes)
Sven:
  75.111 ns (4 allocations: 416 bytes)
Erik:
  4.791 ns (0 allocations: 0 bytes)
```

I also verified that they are all correct:
```julia
@testset verbose=true "triangle plane intersection" begin
    cell_size = (1.0, 1.0, 1.0)
    min_corners = vec(SVector.(Tuple.(CartesianIndices((10, 10, 10)))))
    plane_points = rand(SVector{3}, 10) .* 10
    plane_normals = [2 * rand(SVector{3}) .- 1 for _ in 1:10]

    for point_on_plane in plane_points
        for plane_normal in plane_normals
            for cell_min_corner in min_corners
                cell_max_corner = cell_min_corner .+ cell_size

                default = TrixiParticles.triangle_plane_intersection(point_on_plane, plane_normal,
                                                                     cell_min_corner, cell_max_corner,
                                                                     cell_size)
                sven = TrixiParticles.triangle_plane_intersection_sven(point_on_plane, plane_normal,
                                                                       cell_min_corner, cell_max_corner,
                                                                       cell_size)
                erik = TrixiParticles.triangle_plane_intersection_erik(point_on_plane, plane_normal,
                                                                       cell_min_corner, cell_max_corner,
                                                                       cell_size)

                @test default == sven == erik
            end
        end
    end
end
```

Reply:
ok I see I should have avoided creating the lists^^

Reply:
The trend still holds when I make all your vectors generators to avoid allocations:
```
Default:
  7.333 ns (0 allocations: 0 bytes)
Sven:
  26.564 ns (0 allocations: 0 bytes)
Erik:
  4.833 ns (0 allocations: 0 bytes)

Default:
  7.333 ns (0 allocations: 0 bytes)
Sven:
  26.538 ns (0 allocations: 0 bytes)
Erik:
  4.833 ns (0 allocations: 0 bytes)

Default:
  7.333 ns (0 allocations: 0 bytes)
Sven:
  26.690 ns (0 allocations: 0 bytes)
Erik:
  4.833 ns (0 allocations: 0 bytes)

Default:
  7.333 ns (0 allocations: 0 bytes)
Sven:
  26.564 ns (0 allocations: 0 bytes)
Erik:
  4.833 ns (0 allocations: 0 bytes)

Default:
  7.333 ns (0 allocations: 0 bytes)
Sven:
  26.564 ns (0 allocations: 0 bytes)
Erik:
  4.833 ns (0 allocations: 0 bytes)

Default:
  7.333 ns (0 allocations: 0 bytes)
Sven:
  26.591 ns (0 allocations: 0 bytes)
Erik:
  4.834 ns (0 allocations: 0 bytes)

Default:
  3.625 ns (0 allocations: 0 bytes)
Sven:
  26.564 ns (0 allocations: 0 bytes)
Erik:
  4.834 ns (0 allocations: 0 bytes)

Default:
  7.333 ns (0 allocations: 0 bytes)
Sven:
  26.564 ns (0 allocations: 0 bytes)
Erik:
  4.833 ns (0 allocations: 0 bytes)

Default:
  7.333 ns (0 allocations: 0 bytes)
Sven:
  26.553 ns (0 allocations: 0 bytes)
Erik:
  4.833 ns (0 allocations: 0 bytes)

Default:
  7.333 ns (0 allocations: 0 bytes)
Sven:
  26.580 ns (0 allocations: 0 bytes)
Erik:
  4.875 ns (0 allocations: 0 bytes)
```

Reply:
It improved by ~500% though :D

Reply:
cool! Thanks a lot